pvlib/modelchain.py

```diff
@@ -810,32 +810,101 @@ def run_model(self, times=None, weather=None):
+            'H', for effective irradiance ratio and effective temperature
+            ratio, respectively. See :py:func:`~pvlib.pvsystem.sdm_campanelli`.
+            If either 'F' or 'H' is missing, then convential weather data are
+            also required to compute the missing 'F' or 'H'.
 
         Returns
         -------
         self
 
-        Assigns attributes: times, solar_position, airmass, irradiance,
-        total_irrad, effective_irradiance, weather, temps, aoi,
-        aoi_modifier, spectral_modifier, dc, ac, losses.
+        For conventional weather data, assigns attributes: times,
+        solar_position, airmass, irradiance, total_irrad, effective_irradiance,
+        weather, temps, aoi, aoi_modifier, spectral_modifier, dc, ac, losses.
+        For complete reference device data, only assigns attributes: times, dc,
+        ac, losses.
         """
 
-        self.prepare_inputs(times, weather)
-        self.aoi_model()
-        self.spectral_model()
-        self.effective_irradiance_model()
-        self.temp_model()
+        if weather is not None and 'F' in weather:
+            # Use PV reference device weather information for F
+            self.F = weather['F']
+        else:
+            # Clear any stale value
+            if hasattr(self, 'F'):
+                print('F cleared')
+                del self.F
+        if weather is not None and 'H' in weather:
+            # Use PV reference device weather information for H
+            self.H = weather['H']
+        else:
+            # Clear any stale value
+            if hasattr(self, 'H'):
+                print('H cleared')
+                del self.H
+        if not hasattr(self, 'F') and not hasattr(self, 'H'):
+            # Use conventional (meterological) weather information
+            self.prepare_inputs(times, weather)
+            self.aoi_model()
+            self.spectral_model()
+            self.effective_irradiance_model()
+            self.temp_model()
         self.dc_model()
         self.ac_model()
         self.losses_model()
 
         return self
+
+
+def sdm_campanelli(mc):
+    """
+    ModelChain wrapper for Campanelli et al. model, optionally using SAPM for
+    effective irradiance F and/or effecitve temperature H.
+
+    TODO Details
+    """
+
+    if not hasattr(mc, 'F') or not hasattr(mc, 'H'):
+        # SAPM computes F and H so one/both may be used
+        F, H = get_F_H_from_sapm(mc.effective_irradiance, mc.temps,
+                                 mc.system.module_parameters['Aisc'],
+                                 mc.system.module_parameters['irrad_ref'],
+                                 mc.system.module_parameters['temp_ref'])
+        if not hasattr(mc, 'F'):
+            mc.F = F
+        if not hasattr(mc, 'H'):
+            mc.H = H
+
+    # Calculate the dc power and assign it to mc.dc
+    mc.dc = pvsystem.sdm_campanelli(mc.F, mc.H, **mc.system.module_parameters)
+
```

Comment:
Oops. I have forgotten to scale the output for `modules_per_string` and `strings_per_inverter`.

Reply:
Can `sdm_campanellie_sapm` be built into `sdm_campanelli`?  The `_sapm` variation only fills in `F` and `H` when they are not supplied as weather data.